File: 2015/day-19/day-19-medicine-for-rudolph.js
const fs = require("fs");
const data = fs.readFileSync(__dirname + "/day-19-data.txt", "utf-8").split("\n");

const { input, replacements } = parseInput(data);

console.log("Part 1: ", makeAllReplacements({ replacements, input }).size);

function parseInput(data) {
  const input = data.pop();
  data.pop();
  const replacements = data.map((ea) => {
    const [from, to] = ea.split(" => ");
    return {
      from,
      to,
    };
  });
  return {
    input,
    replacements,
  };
}

function makeAllReplacements({ input, replacements }) {
  const uniqueStrings = new Set();
  replacements
    .map((replacement) => {
      const newStrings = [];
      for (let i = 0; i < input.length; i++) {
        if (input.slice(i, i + replacement.from.length) === replacement.from) {
          const newString = input.slice(0, i) + replacement.to + input.slice(i + replacement.from.length);
          newStrings.push(newString);
        }
      }
      return newStrings;
    })
    .flat()
    .forEach((string) => {
      uniqueStrings.add(string);
    });
  return uniqueStrings;
}

const numberOfMolecules = input.match(/[A-Z]/g).length;
const exteriorCantReplace = input.match(/Ar|Rn/g).length;
const interiorCantReplace = input.match(/Y/g).length;
console.log("Part 2: ", numberOfMolecules - exteriorCantReplace - (2 * interiorCantReplace) - "e".length);

File: 2015/day-19/day-19-medicine-for-rudolph.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
)

func main() {
	f, _ := os.Open("./day-19-data.txt")
	s := bufio.NewScanner(f)
	rawInstructions := []string{}
	for s.Scan() {
		rawInstructions = append(rawInstructions, s.Text())
	}

	replacements, molecule := rawInstructions[:len(rawInstructions)-2], rawInstructions[len(rawInstructions)-1]

	replacementMap := make(map[string][]string)
	for _, r := range replacements {
		from, to, _ := strings.Cut(r, " => ")
		_, exists := replacementMap[from]
		if exists {
			replacementMap[from] = append(replacementMap[from], to)
		} else {
			replacementMap[from] = []string{to}
		}
	}

	moleculeSet := make(map[string]struct{})
	for k, v := range replacementMap {
		for i := 0; i < len(molecule)-(len(k)-1); i++ {
			sub := molecule[i : i+len(k)]
			if sub == k {
				for _, val := range v {
					replaced := molecule[:i] + val + molecule[i+len(k):]
					moleculeSet[replaced] = struct{}{}
				}
			}
		}
	}

	fmt.Println("Part one:", len(moleculeSet))

	r := regexp.MustCompile("[A-Z]")
	numberOfMolecules := len(r.Split(molecule, -1)) - 1
	r = regexp.MustCompile("Ar|Rn")
	exteriorCantReplace := len(r.Split(molecule, -1)) - 1
	r = regexp.MustCompile("Y")
	interiorCantReplace := len(r.Split(molecule, -1)) - 1

	fmt.Println("Part two:", numberOfMolecules-exteriorCantReplace-(2*interiorCantReplace)-1)
}
